stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -10,7 +10,7 @@ struct Diagonal{T,V<:AbstractVector{T}} <: AbstractMatrix{T}
         new{T,V}(diag)
     end
 end
-Diagonal{T,V}(d::Diagonal) where {T,V<:AbstractVector{T}} = Diagonal{T,V}(d.diag)
+Diagonal{T,V}(M::AbstractMatrix) where {T,V<:AbstractVector{T}} = Diagonal{T,V}(diag(M))
```

Comment:
Should this be moved then to line 105?

Reply:
We might want to have the constructor anyway, and retain the general behavior of `convert` calling the constructor. The difference here is that `convert` ascertains that the argument is diagonal, whereas the constructor accepts any matrix and extracts the principal diagonal. This matches the other constructors that already exist.

Reply:
I meant just copy-paste there, not to replace anything. Because the `Diagonal***(::AbstractMatrix)` methods are listed there, this one was odd here even before your PR.